src/ggml-impl.h

```diff
@@ -311,29 +311,24 @@ GGML_API void ggml_aligned_free(void * ptr, size_t size);
 
 // FP16 to FP32 conversion
 
+// 16-bit float
+// on Arm, we use __fp16
+// on x86, we use uint16_t
 #if defined(__ARM_NEON)
```

Comment:
Shouldn't there be a corresponding ` && (!defined(__MSC_VER)` here? Since we're only removing the CUDA check

Reply:
I think if MSVC supports `__fp16`, we don't need this check. But maybe it was added because it does not support it. If that's the case, we'll have to bring it back and fallback to the reference implementation.

Reply:
> I think if MSVC supports __fp16, we don't need this check.

Yeah true. The original PR said: `can't use native __fp16 type as it's not supported by MSVC` - https://github.com/ggml-org/llama.cpp/pull/3007

Maybe it has changed since that PR, but I can't find anything to this on the internet though.

Reply:
Aha, this explains it.

But on the other hand, we have a MSVC+Arm CI job in `llama.cpp` that appears to be passing with this change:

https://github.com/ggml-org/llama.cpp/actions/runs/14262095042/job/39975786169

So maybe it was indeed fixed?

Reply:
Sounds good. This could be merged in, and if there are any reports of regression on MSVC+Arm Neon, then it's easy to fix it back.

Reply:
> MSVC cannot be used to build for Arm, the CPU backend does not allow it.

This is because of the issue with the Arm inline assembly that is not supported by the MSVC compiler, correct?

Reply:
Support for Arm MSVC was removed when fixing the CPU feature check with `GGML_NATIVE` in cmake, since it would require writing a specific implementation for it. You could still build with MSVC without the inline assembly kernels before that, it would just be slower. Since Clang is always available and shipped with Visual Studio, there was no reason to spend time on that.

Reply:
Got it. So I think it that case, this PR should be good to merge?

Reply:
I did some additional digging and found the need for the CUDA and MUSA checks. They do seem to be necessary so I added comments in the code and updated the OP to try to make things a bit more clear. This PR now just removes the `ggml_fp16_internal_t` type which is now redundant and cleans up some macros and include logic.

Let me know if you see any red flags.